File: src/audio/diarizer.ts
/**
 * diarizer.ts — TypeScript wrapper for diarizer.py
 *
 * Spawns the Python diarization process and exposes a simple async API.
 * Used by VAD to tag each audio chunk with speaker identity before
 * passing to the decision pipeline.
 *
 * Speaker labels:
 *   'self'    — Tanay speaking (ARIA should NOT fire)
 *   'other'   — other person speaking (ARIA should fire if signal detected)
 *   'unknown' — couldn't determine (treat as 'other' to avoid missing signals)
 *
 * Usage:
 *   import { getDiarizer } from './diarizer.js'
 *   const d = getDiarizer()
 *   const result = await d.identify(audioBuffer)
 *   if (result.speaker !== 'self') { ... }
 */

import { spawn, ChildProcess } from 'child_process'
import * as path from 'path'
import { fileURLToPath } from 'url'
import { CONFIG } from '../config.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname  = path.dirname(__filename)
const SCRIPT     = path.resolve(__dirname, '..', '..', 'src', 'audio', 'diarizer.py')

export type SpeakerLabel = 'self' | 'other' | 'unknown'

export interface DiarizationResult {
  speaker:    SpeakerLabel
  confidence: number
  reason?:    string
}

export interface DiarizerStatus {
  enrolled:             boolean
  profile_exists:       boolean
  enroll_progress_pct:  number
}

// ── Pending request queue ──────────────────────────────────────────────────
// Each request is a resolve/reject pair waiting for the next stdout line.

interface PendingRequest {
  resolve: (result: DiarizationResult) => void
  reject:  (err: Error) => void
  timeout: NodeJS.Timeout
}

class Diarizer {
  private proc:    ChildProcess | null = null
  private queue:   PendingRequest[]    = []
  private buffer:  string              = ''
  private ready:   boolean             = false
  private enabled: boolean             = true

  constructor() {
    if (process.env.NODE_ENV === 'test' || process.env.NO_DIARIZER === '1') {
      this.enabled = false
      console.log('[DIARIZER] disabled (test/env)')
      return
    }
    this.spawn()
  }

  private spawn() {
    try {
      this.proc = spawn(CONFIG.VENV_PYTHON, [SCRIPT])

      this.proc.stdout?.on('data', (chunk: Buffer) => {
        this.buffer += chunk.toString()
        const lines = this.buffer.split('\n')
        this.buffer  = lines.pop() ?? ''

        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed) continue

          // Ready signal
          if (trimmed === '[DIARIZER] ready') {
            this.ready = true
            console.log('[DIARIZER] ready')
            continue
          }

          // Try to parse as JSON result
          try {
            const parsed = JSON.parse(trimmed)
            const pending = this.queue.shift()
            if (pending) {
              clearTimeout(pending.timeout)
              pending.resolve(parsed as DiarizationResult)
            }
          } catch {
            // Log line (status/progress messages)
            console.log(trimmed)
          }
        }
      })

      this.proc.stderr?.on('data', (d: Buffer) => {
        const msg = d.toString().trim()
        if (msg && !msg.startsWith('UserWarning') && !msg.startsWith('torch')) {
          console.error('[DIARIZER]', msg)
        }
      })

      this.proc.on('exit', (code) => {
        console.log(`[DIARIZER] process exited (code=${code})`)
        this.proc  = null
        this.ready = false
        // Drain pending with unknown
        for (const p of this.queue) {
          clearTimeout(p.timeout)
          p.resolve({ speaker: 'unknown', confidence: 0, reason: 'process_exited' })
        }
        this.queue = []
        // Respawn after 2s
        setTimeout(() => this.spawn(), 2000)
      })

      console.log('[DIARIZER] spawned')
    } catch (e) {
      console.error('[DIARIZER] failed to spawn:', e)
      this.enabled = false
    }
  }

  // ── Public API ─────────────────────────────────────────────────────────

  async identify(audio: Buffer, timeoutMs = 800): Promise<DiarizationResult> {
    if (!this.enabled || !this.proc) {
      return { speaker: 'unknown', confidence: 0, reason: 'disabled' }
    }

    const audio_b64 = audio.toString('base64')
    const payload   = JSON.stringify({ audio_b64 }) + '\n'

    return new Promise((resolve, reject) => {
      const timeout = setTimeout(() => {
        // Remove from queue
        this.queue = this.queue.filter(p => p.resolve !== resolve)
        // Default to 'other' on timeout so we don't miss signals
        resolve({ speaker: 'other', confidence: 0.5, reason: 'timeout' })
      }, timeoutMs)

      this.queue.push({ resolve, reject, timeout })

      try {
        this.proc?.stdin?.write(payload)
      } catch (e) {
        clearTimeout(timeout)
        this.queue = this.queue.filter(p => p.resolve !== resolve)
        resolve({ speaker: 'unknown', confidence: 0, reason: 'write_error' })
      }
    })
  }

  async reenroll(): Promise<void> {
    if (!this.proc) return
    this.proc.stdin?.write(JSON.stringify({ cmd: 'reenroll' }) + '\n')
    console.log('[DIARIZER] re-enrollment triggered')
  }

  async getStatus(): Promise<DiarizerStatus | null> {
    if (!this.proc) return null

    return new Promise((resolve) => {
      const timeout = setTimeout(() => resolve(null), 1000)

      // Status response comes back as a JSON line too — intercept next
      const originalResolve = (result: DiarizationResult) => {
        clearTimeout(timeout)
        resolve(result as unknown as DiarizerStatus)
      }

      this.queue.push({ resolve: originalResolve, reject: () => resolve(null), timeout })
      this.proc?.stdin?.write(JSON.stringify({ cmd: 'status' }) + '\n')
    })
  }

  isReady(): boolean {
    return this.ready && this.enabled
  }

  isEnabled(): boolean {
    return this.enabled
  }
}

// ── Singleton ──────────────────────────────────────────────────────────────

let _instance: Diarizer | null = null

export function getDiarizer(): Diarizer {
  if (!_instance) _instance = new Diarizer()
  return _instance
}

// Pre-warm on module load
getDiarizer()

File: src/audio/diarizer.py
#!/usr/bin/env python3
"""
diarizer.py — Real-time speaker diarization for ARIA
=====================================================

Identifies who is speaking: 'self' (Tanay) or 'other' (prospect/person).

Strategy:
  - Uses resemblyzer locally (no cloud, no pyannote license needed)
  - On first run, enrolls a voice profile for 'self' from a short reference clip
  - Each audio chunk is diarized + speaker-matched against the enrolled profile
  - Outputs JSON: {"speaker": "self"|"other"|"unknown", "confidence": 0.0-1.0}

Voice enrollment:
  - On startup, looks for ~/A.R.I.A/voice_profile.wav
  - If not found, first 5s of speech is used as self-enrollment (bootstrap mode)
  - Can be re-enrolled by deleting the file and restarting

Dependencies:
  pip install resemblyzer torch --break-system-packages

Usage (standalone):
  echo '{"audio_b64": "<base64 pcm>"}' | python3 diarizer.py

Usage (from Node via stdin/stdout pipe):
  Spawned by diarizer.ts, communicates via JSON lines on stdin/stdout
"""

from __future__ import annotations

import sys
import json
import os
import base64
import io
import struct
import wave
from typing import Optional

import numpy as np  # type: ignore[import-not-found]

# ── Config ─────────────────────────────────────────────────────────────────

SAMPLE_RATE          = 16000
CHANNELS             = 1
PROFILE_PATH         = os.path.expanduser("~/A.R.I.A/voice_profile.wav")
ARIA_DIR             = os.path.expanduser("~/A.R.I.A")
ENROLL_SECONDS       = 5      # seconds of speech needed for self-enrollment
MIN_CHUNK_MS         = 500    # ignore chunks shorter than this
SIMILARITY_THRESHOLD = 0.75   # cosine similarity to consider "self"

# ── Type-safe globals ──────────────────────────────────────────────────────
# encoder and self_embedding start as None and are set after load_encoder().
# All call-sites guard against None before use.

_encoder: Optional[object] = None          # resemblyzer.VoiceEncoder once loaded
_self_embedding: Optional[np.ndarray] = None
_enroll_buffer: list[bytes] = []
_enroll_duration_ms: float = 0.0
_enrolled: bool = False

# ── Model loading ──────────────────────────────────────────────────────────

def load_encoder() -> bool:
    global _encoder
    try:
        from resemblyzer import VoiceEncoder  # type: ignore[import-untyped]
        _encoder = VoiceEncoder()
        print("[DIARIZER] VoiceEncoder loaded", flush=True)
        return True
    except Exception as e:
        print(f"[DIARIZER] VoiceEncoder unavailable: {e}", flush=True)
        return False


def load_self_profile() -> bool:
    global _self_embedding, _enrolled
    if not os.path.exists(PROFILE_PATH):
        print("[DIARIZER] No voice profile found — will auto-enroll from first speech", flush=True)
        return False
    if _encoder is None:
        return False
    try:
        from resemblyzer import preprocess_wav  # type: ignore[import-untyped]
        wav = preprocess_wav(PROFILE_PATH)
        _self_embedding = _encoder.embed_utterance(wav)  # type: ignore[union-attr]
        _enrolled = True
        print("[DIARIZER] Voice profile loaded", flush=True)
        return True
    except Exception as e:
        print(f"[DIARIZER] Failed to load profile: {e}", flush=True)
        return False

# ── PCM helpers ────────────────────────────────────────────────────────────

def pcm_to_wav_bytes(pcm_bytes: bytes) -> bytes:
    """Wrap raw 16-bit PCM in a WAV container."""
    buf = io.BytesIO()
    with wave.open(buf, "wb") as wf:
        wf.setnchannels(CHANNELS)
        wf.setsampwidth(2)  # 16-bit
        wf.setframerate(SAMPLE_RATE)
        wf.writeframes(pcm_bytes)
    return buf.getvalue()


def duration_ms(pcm_bytes: bytes) -> float:
    return (len(pcm_bytes) / 2 / SAMPLE_RATE) * 1000

# ── Enrollment ─────────────────────────────────────────────────────────────

def try_enroll(pcm_bytes: bytes) -> bool:
    """
    Accumulate audio until we have ENROLL_SECONDS of speech,
    then save as voice profile and embed.
    Returns True when enrollment is complete.
    """
    global _enroll_buffer, _enroll_duration_ms, _self_embedding, _enrolled

    if _encoder is None:
        return False

    _enroll_buffer.append(pcm_bytes)
    _enroll_duration_ms += duration_ms(pcm_bytes)

    if _enroll_duration_ms < ENROLL_SECONDS * 1000:
        pct = int(_enroll_duration_ms / (ENROLL_SECONDS * 1000) * 100)
        print(f"[DIARIZER] Enrolling self... {pct}%", flush=True)
        return False

    # Enough audio — save and embed
    combined = b"".join(_enroll_buffer)
    os.makedirs(ARIA_DIR, exist_ok=True)

    with open(PROFILE_PATH, "wb") as f:
        f.write(pcm_to_wav_bytes(combined))

    try:
        from resemblyzer import preprocess_wav  # type: ignore[import-untyped]
        wav = preprocess_wav(PROFILE_PATH)
        _self_embedding = _encoder.embed_utterance(wav)  # type: ignore[union-attr]
        _enrolled = True
        print("[DIARIZER] Self-enrollment complete — voice profile saved", flush=True)
        return True
    except Exception as e:
        print(f"[DIARIZER] Enrollment failed: {e}", flush=True)
        return False

# ── Cosine similarity ──────────────────────────────────────────────────────

def cosine_sim(a: np.ndarray, b: np.ndarray) -> float:
    norm_a = float(np.linalg.norm(a))
    norm_b = float(np.linalg.norm(b))
    if norm_a == 0.0 or norm_b == 0.0:
        return 0.0
    return float(np.dot(a, b) / (norm_a * norm_b))

# ── Diarize a single chunk ─────────────────────────────────────────────────

def diarize_chunk(pcm_bytes: bytes) -> dict:
    """
    Returns {"speaker": "self"|"other"|"unknown", "confidence": float}
    """
    if _encoder is None:
        return {"speaker": "unknown", "confidence": 0.0, "reason": "encoder_not_loaded"}

    dur = duration_ms(pcm_bytes)
    if dur < MIN_CHUNK_MS:
        return {"speaker": "unknown", "confidence": 0.0, "reason": "chunk_too_short"}

    # Bootstrap enrollment — treat first chunks as "self"
    if not _enrolled:
        complete = try_enroll(pcm_bytes)
        if not complete:
            return {"speaker": "self", "confidence": 0.5, "reason": "enrolling"}
        return {"speaker": "self", "confidence": 0.9, "reason": "enrolled"}

    # self_embedding is guaranteed non-None here because _enrolled is True
    assert _self_embedding is not None

    # Embed the incoming chunk
    try:
        from resemblyzer import preprocess_wav  # type: ignore[import-untyped]

        wav_bytes = pcm_to_wav_bytes(pcm_bytes)
        wav = preprocess_wav(io.BytesIO(wav_bytes))

        if len(wav) < SAMPLE_RATE * 0.3:
            return {"speaker": "unknown", "confidence": 0.0, "reason": "too_short_after_preprocess"}

        chunk_embedding: np.ndarray = _encoder.embed_utterance(wav)  # type: ignore[union-attr]
        similarity = cosine_sim(chunk_embedding, _self_embedding)

        if similarity >= SIMILARITY_THRESHOLD:
            return {
                "speaker": "self",
                "confidence": round(similarity, 3),
                "reason": "similarity_match",
            }
        else:
            return {
                "speaker": "other",
                "confidence": round(1.0 - similarity, 3),
                "reason": "similarity_mismatch",
            }

    except Exception as e:
        return {"speaker": "unknown", "confidence": 0.0, "reason": str(e)}

# ── Re-enrollment command ──────────────────────────────────────────────────

def handle_reenroll() -> None:
    global _enroll_buffer, _enroll_duration_ms, _self_embedding, _enrolled
    if os.path.exists(PROFILE_PATH):
        os.remove(PROFILE_PATH)
    _enroll_buffer = []
    _enroll_duration_ms = 0.0
    _self_embedding = None
    _enrolled = False
    print("[DIARIZER] Re-enrollment triggered — speak to re-enroll", flush=True)

# ── Main loop ──────────────────────────────────────────────────────────────

def main() -> None:
    print("[DIARIZER] loading encoder...", flush=True)
    ok = load_encoder()
    if not ok:
        print("[DIARIZER] running in fallback mode (no encoder)", flush=True)
    else:
        load_self_profile()

    print("[DIARIZER] ready", flush=True)

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        try:
            msg: dict = json.loads(line)
        except Exception:
            continue

        cmd = msg.get("cmd")

        if cmd == "reenroll":
            handle_reenroll()
            print(json.dumps({"ok": True, "msg": "re-enrollment started"}), flush=True)
            continue

        if cmd == "status":
            print(json.dumps({
                "enrolled": _enrolled,
                "profile_exists": os.path.exists(PROFILE_PATH),
                "enroll_progress_pct": int(
                    min(100.0, _enroll_duration_ms / (ENROLL_SECONDS * 1000) * 100)
                ),
            }), flush=True)
            continue

        # Default: diarize audio chunk
        audio_b64: Optional[str] = msg.get("audio_b64")
        if not audio_b64:
            print(json.dumps({"speaker": "unknown", "confidence": 0.0, "reason": "no_audio"}), flush=True)
            continue

        try:
            pcm_bytes = base64.b64decode(audio_b64)
        except Exception as e:
            print(json.dumps({"speaker": "unknown", "confidence": 0.0, "reason": f"decode_error: {e}"}), flush=True)
            continue

        result = diarize_chunk(pcm_bytes)
        print(json.dumps(result), flush=True)


if __name__ == "__main__":
    main()